alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -0,0 +1,155 @@
+import qualified Plutus.V1.Ledger.Value as P (Value (..))
+import Shelley.Spec.Ledger.Address (Addr (..), BootstrapAddress (..))
+import Shelley.Spec.Ledger.BaseTypes (StrictMaybe (..))
+import Shelley.Spec.Ledger.Coin (Coin (..))
+import Shelley.Spec.Ledger.Credential (Credential (KeyHashObj, ScriptHashObj))
+import Shelley.Spec.Ledger.Keys (KeyHash (..))
+import Shelley.Spec.Ledger.Scripts (ScriptHash (..))
+import Shelley.Spec.Ledger.TxBody (TxId (..), TxIn (..))
+
+-- =========================================================
+
+transSafeHash :: SafeHash c i -> BS.ByteString
+transSafeHash safe = case extractHash safe of UnsafeHash b -> fromShort b
+
+transTxId :: TxId era -> P.TxId
+transTxId (TxId safe) = P.TxId (transSafeHash safe)
+
+transVI :: ValidityInterval -> P.SlotRange
+transVI (ValidityInterval SNothing SNothing) =
+  P.Interval
+    (P.LowerBound P.NegInf True)
+    (P.UpperBound P.PosInf False)
+transVI (ValidityInterval (SJust (SlotNo i)) SNothing) =
+  P.Interval
+    (P.LowerBound (P.Finite (fromIntegral i)) True)
+    (P.UpperBound P.PosInf False)
+transVI (ValidityInterval SNothing (SJust (SlotNo i))) =
+  P.Interval
+    (P.LowerBound P.NegInf True)
+    (P.UpperBound (P.Finite (fromIntegral i)) False)
+transVI (ValidityInterval (SJust (SlotNo i)) (SJust (SlotNo j))) =
+  P.Interval
+    (P.LowerBound (P.Finite (fromIntegral i)) True)
+    (P.UpperBound (P.Finite (fromIntegral j)) False)
+
+transTxIn' :: CC.Crypto c => TxIn c -> P.TxOutRef
+transTxIn' (TxIn txid nat) = P.TxOutRef (transTxId txid) (fromIntegral nat)
+
+transTxIn :: forall c. CC.Crypto c => TxIn c -> P.TxInInfo
+transTxIn txin =
+  P.TxInInfo
+    (transTxIn' txin)
+    Nothing -- NOT SURE ABOUT THIS
+    (transValue (inject @(Mary.Value c) (Coin 0))) -- OR ABOUT THIS
+
+transTxOut :: forall era. (Era era, Value era ~ Mary.Value (Crypto era)) => TxOut era -> P.TxOut
+transTxOut (TxOut (Addr _network (ScriptHashObj sh) _stake) val datahash) =
+  P.TxOut
+    (P.ScriptAddress (transScriptHash sh))
+    (transValue @(Crypto era) val)
+    (P.PayToScript (transDataHash datahash))
+transTxOut (TxOut (Addr _network (KeyHashObj kh) _stake) val datahash) =
+  P.TxOut
+    (P.PubKeyAddress (transKeyHash kh))
+    (transValue @(Crypto era) val)
+    (P.PayToScript (transDataHash datahash))
+transTxOut (TxOut (AddrBootstrap (BootstrapAddress _ad)) val datahash) =
+  P.TxOut
+    (P.PubKeyAddress (P.PubKeyHash undefined)) -- TODO BootstrapAddress into ByteString
+    (transValue @(Crypto era) val)
+    (P.PayToScript (transDataHash datahash))
+
+transValue :: forall c. Mary.Value c -> P.Value
+transValue (Mary.Value _coin _mp) = undefined -- TODO complete this
+
+transKeyHash :: KeyHash d c -> P.PubKeyHash
+transKeyHash (KeyHash (UnsafeHash h)) = P.PubKeyHash (fromShort h)
+
+transScriptHash :: ScriptHash i -> P.ValidatorHash
+transScriptHash (ScriptHash (UnsafeHash h)) = P.ValidatorHash (fromShort h)
+
+transDataHash :: StrictMaybe (DataHash c) -> P.DatumHash
```

Comment:
I think we should follow your naming for `KeyHash`, `ScriptHash`... but I think it should be `DatumHash` on your side?

Reply:
left it for now